### DISCUSSION TITLE
Model is not learning, model parameters require grad not grad is None

### DISCUSSION THREAD

**Discussion by zzzLemon** [pyg-team/pytorch_geometric#6250] [Q&A]
Hi,

I'm working on heterogenous graphs training, and here is the previous post 
[https://github.com/pyg-team/pytorch_geometric/discussions/6191](url)

Since the loss remains nearly constant, I print the model parameters after loss.backward() of each batch of the model using model.namedparameters(), and they are constant during training, and here is one parameter:

```
HeteroGNN_layers.0.convs.transistor__transistor_connect_to_transistor__transistor.lin_l.weight tensor(
        [[ 0.1008, -0.2823,  0.5347],
        [ 0.1817, -0.3962,  0.2257],
        [-0.1071, -0.1242, -0.4817],
        [-0.4998,  0.1904,  0.0080],
        [-0.2331, -0.0823,  0.3008],
        [ 0.3688, -0.2635,  0.5177],
        [ 0.2579,  0.5483, -0.0702],
        [ 0.3688, -0.3630, -0.0988],
        [-0.1795, -0.3418,  0.1507],
        [ 0.5689,  0.3813, -0.2845]], device='cuda:2') Batch_index : 0
HeteroGNN_layers.0.convs.transistor__transistor_connect_to_transistor__transistor.lin_l.weight tensor(
        [[ 0.1008, -0.2823,  0.5347],
        [ 0.1817, -0.3962,  0.2257],
        [-0.1071, -0.1242, -0.4817],
        [-0.4998,  0.1904,  0.0080],
        [-0.2331, -0.0823,  0.3008],
        [ 0.3688, -0.2635,  0.5177],
        [ 0.2579,  0.5483, -0.0702],
        [ 0.3688, -0.3630, -0.0988],
        [-0.1795, -0.3418,  0.1507],
        [ 0.5689,  0.3813, -0.2845]], device='cuda:2') Batch_index : 1
```
And then I check the grad of this parameter, its None...
I'm wondering how can this happen?  I'm following the heteroConv tutorial [https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html](url), and here is my forward function.
```
    def forward(self, x_in):
        for conv in self.HeteroGNN_layers:
            #x_dict = conv(x_dict, edge_index_dict)
            x_dict = conv(x_in.x_dict, x_in.edge_index_dict)
            x_dict = {key: F.relu(x) for key, x in x_dict.items()}
        ptr=   x_in["wire"].ptr
        ptr= ptr[1:] -1
        x_out=x_dict['wire'][ptr]
        x_out=torch.sigmoid(self.linear(x_out))
        return (x_out)# the last wire node

```
x_in is a batch. I think the reason could be, x_in is batch and cant be forwarded. So I have modified the train() function into:
```
 batch=batch.to('cuda:2')
            optimizer.zero_grad()
            out=model(batch.x_dict,batch.edge_index_dict)
```
And I get error " 'Tensor' object has no attribute 'ptr' "", because need ptr of a node, which is only possible to do when I use batch as the argument of forward. I would appreciate it if someone could help, thanks

**Comment by rusty1s** [comment]
How does your training loop look like? If you have no gradients, then did you forget to call `loss.backward()`? Note that you should also add `ptr` to your argument list of the model, e.g., model(batch.x_dict, batch.edge_index_dict, batch["wire"].ptr)`

**Reply by zzzLemon** [reply]
Hi, thanks for your reply. I also noticed the ptr should be an argument. 

```
 for epoch in range(num_epochs):
        model.to(device)
        model.train()
        for batch in data_loader:
            BAT_transistor=torch.nn.BatchNorm1d(3)
            BAT_wire=torch.nn.BatchNorm1d(1)
            batch.x_dict["transistor"]=BAT_transistor(batch.x_dict["transistor"])
            batch.x_dict["wire"]=BAT_wire(batch.x_dict["wire"])
            batch=batch.to(device)
            optimizer.zero_grad()      
            out=model(batch.x_dict,batch.edge_index_dict,batch["wire"].ptr)
            y_predict = batch.y.view(out.shape)
            loss =torch.nn.MSELoss()(y_predict,out)
            batch_loss.append
            loss.backward()
            optimizer.step()
            train_loss += loss.item()*batch_size
```
However, I made some tests yesterday, although the weights/bias are updated, the finial loss has nearly no difference with the loss when I just do 
```
out=self.relu(self.linear(x_dict['wire'][batch["wire"].ptr[1:]-1]))
```
without any heterconv layer. What the reason could be? Currently, I have 5 layers and out_channels for each layer are set to 8, and in forward() of the model with heterconv, it's like:
```    
def forward(self, x_dict,edge_index,batch_ptr):
        x_dict = self.conv1(x_dict, edge_index)
        x_dict = {key: self.relu(x) for key, x in x_dict.items()}
        x_dict = self.conv2(x_dict, edge_index)
        x_dict = {key: self.relu(x) for key, x in x_dict.items()}
        ***** three more conv layers here
        x_out=self.linear(x_dict['wire'][batch_ptr[1:]-1])
```
I think the code itself has no bug, does it mean my dataset is inherently bad and the graph neural network cant model the problem very well?

**Reply by zzzLemon** [reply]
I checked some discussions like [https://github.com/pyg-team/pytorch_geometric/discussions/4140](url). I did some tests this morning, here is the loss/epoch plot. The size of conv (SAGEConv) was fixed: 4 layers, hidden channel 10. Black line: conv out layer connects to a linear layer 
```    
        ******
        x_dict = self.conv4(x_dict, edge_index)
        x_out=self.linear(x_dict['wire'][batch_ptr[1:]-1])
```
Red line, there is no graph neural network, I simply do: 
```
x_out=self.linear(x_dict['wire'][batch_ptr[1:]-1])
```
Yellow line, I check the discussion in the link, and remove the linear layer, so the last conv layer is:

```      
         *******
        self.conv4= HeteroConv({
        ('input', 'input_connect_to_wire', 'wire'): SAGEConv((-1,-1), 1),

```

And green line shows the result after I add batchnorm1d to each conv layer, also the linear layer is also removed. Blue line, I connect another mlp to the last conv layer.
You see they have nearly no difference. So I'm doubting whether the network is learning or whether there is something wrong with the dataset


![loss_epoch](https://user-images.githubusercontent.com/40807005/208293478-0ff5722c-b01e-4b39-a49b-4e796c6d812c.PNG)

**Reply by rusty1s** [reply]
Interesting. Is your node you use for final readout connected to the underlying graph? Is this binary classification? What happens if you use binary cross entropy instead of MSE loss?

**Reply by zzzLemon** [reply]
> underlying graph

it's an underlying graph, and it's regression. I tried yesterday to regenerate my dataset and used part of it to train, also increased learning rate, and the loss reduced to 0.75. So I think there are many same structures in my dataset and somehow the model struggled with it. 
Another interesting thing I observed yesterday is when I increased the out_channel of each layer, i.e. I have different types of node in a graph, each node can have maximal 3 features, after I increased the out_channel number to 10, the loss was higher much than when I set the out_channel number to 3. How should this out_channel be set?

**Reply by zzzLemon** [reply]
> Interesting. Is your node you use for final readout connected to the underlying graph? Is this binary classification? What happens if you use binary cross entropy instead of MSE loss?

Hi, I change the activation function from leakyrelu to relu, and the optimizer from sgd to adam, the loss reduces to 0.04. It seems like GAT is not well compatible with leakyrelu, I don't understand why, and I will take a look into it. Anyway, many thanks for your answer.
